src/applications/apps.ts

```diff
@@ -383,75 +411,95 @@ function validCustomProps(customProps) {
   );
 }
 
-function sanitizeArguments(
-  appNameOrConfig,
-  appOrLoadApp,
-  activeWhen,
-  customProps
-) {
+interface ApplicationRegistration {
+  name: string;
+  loadApp: Application;
+  activeWhen: ActivityFn;
+  customProps?: CustomProps;
+}
+
+function sanitizeArguments<ExtraProps extends CustomProps = {}>(
+  appNameOrConfig: string | RegisterApplicationConfig,
+  appOrLoadApp: Application,
+  activeWhen: Activity,
+  customProps?: ExtraProps | CustomPropsFn<ExtraProps>
+): ApplicationRegistration {
   const usingObjectAPI = typeof appNameOrConfig === "object";
 
-  const registration = {
-    name: null,
-    loadApp: null,
-    activeWhen: null,
-    customProps: null,
+  let unsanitizedRegistration: {
```

Comment:
This is a small refactor - the `registration` object used to start off unsanitized and then become sanitized later. To facilitate better type checking, I split it into two objects (`unsanitizedRegistration` and then the returned object`